usr/man/man4/rp.4
.TH RP 4 
.SH NAME
rp \- RP-11/RP03 moving-head disk
.SH DESCRIPTION
The files
.I "rp0 ... rp7"
refer to sections of RP disk drive 0.
The files
.I "rp8 ... rp15"
refer to drive 1 etc.
This is done since the size of
a full RP drive is 81200 blocks and
internally the system is only capable
of addressing 65536
blocks.
Also since the disk is so large,
this allows it to be broken up
into more manageable pieces.
.PP
The origin and size of the pseudo-disks on each drive are
as follows:
.PP
.br
	disk	start	length
.br
	0	0	40600
.br
	1	40600	40600
.br
	2	0	9200
.br
	3	72000	9200
.br
	4	0	65535
.br
	5	15600	65535
.br
	6-7	unassigned
.PP
It is unwise for all of these files to be present in one installation,
since there is overlap in addresses and protection becomes
a sticky matter.
Here is a suggestion for two useful configurations:
If the root of the file system is on some other device
and the RP used as a mounted device,
then
.I rp0
and
.I rp1,
which divide the disk into two equal size portions, is a good idea.
Other things being equal, it is advantageous to have two equal-sized portions
since one can always be copied onto the other, which is occasionally useful.
.PP
If the RP is the only disk and has to
contain the root and the swap area,
the root can be put on
.I rp2
and a mountable file system on
.IR rp5 .
Then the swap space can be put in the
unused blocks 9200 to 15600 of
.I rp0
(or, equivalently,
.IR rp4) .
This arrangement puts the root file system, the swap area,
and the i-list of the mounted file system relatively
near each other and thus tends to
minimize head movement.
.PP
The
.I rp
files
access the disk via the system's normal
buffering mechanism
and may be read and written without regard to
physical disk records.
There is also a `raw' interface
which provides for direct transmission between the disk
and the user's read or write buffer.
A single read or write call results in exactly one I/O operation
and therefore raw I/O is considerably more efficient when
many words are transmitted.
The names of the raw RP files
begin with
.I rrp
and end with a number which selects the same disk
section as the corresponding
.I rp
file.
.PP
In raw I/O the buffer must begin on a word boundary,
and counts should be a multiple of 512 bytes
(a disk block).
Likewise
.I seek
calls should specify a multiple of 512 bytes.
.SH FILES
/dev/rp?, /dev/rrp?
.SH SEE ALSO
hp(4)


named/CHANGES
This is version 4.8.1 of bind.  It is a test release that includes
version 4.8 with fixes, asynchronous zone transfer and better reload
capabilities.  Although it is currently described as a test release,
it is believed to be reasonably stable and more usable than the currently-
released version, 4.8.  The changes of note are:

  o The asynchronous zone transfer code previously posted to the bind
    mailing list has been integrated, completed and tested.  There are
    a number of changes from the version posted, including fixes to
    allow top-level domains to work and a simplification of the timer
    code.

  o The code for reloading the server has been changed so that only
    primary zones master files that have changed are reloaded.  The
    cache and secondary zones are not flushed, and the sortlist, domain,
    etc. are reset to correspond to the boot file contents.

  o Several bugs have been fixed: the name "*" is not interpreted as
    a wildcard in cached zones, only in primary zones.  Secondary servers
    no longer decrement the time-to-live of records by the time since
    they verified the zone with the master; as a result, they never
    hand out nameserver referrals with too short a TTL to be usable.
    A bug was fixed that caused secondary servers with out-of-date
    zones to return empty answers between the actual expiration time
    and the next timeout.

There are several other bugs that have been reported but have not yet
been fixed.  In addition, the next regular release of named will
support negative caching, but this has not been integrated.

I would appreciate receiving feedback on this release; in particular,
problems (or lack of problems) when installing on various systems.
I attempted to update the SysV code when integrating the zone-transfer,
but haven't tested it.

The notes from version 4.8 follow.

----------
Welcome to version 4.8 of bind.

There have been several changes to the named boot file (/etc/named.boot)
of which you should be aware.  The "domain" line for each zone is no longer
needed, but one such line may still be used to specify a default domain
to be used for queries containing names with only a single component.
The term "suffixes", which was added in version 4.7alpha, has been removed.

The manual page on named (named.8) has been updated to reflect all
these changes.  Please read this and look at the example files
before installation.  You should also note the changes in the
resolver code to support non-fully-qualified addresses and per-user
host aliases.  See hostname(7) for an overview.  Two new routines
have been added to the resolver library since the last test release:
res_query formulates a query, sends it, waits for a response and does
preliminary error checking; res_search implements the search rules
of gethostbyname using res_query.

The MX lookup routine in sendmail has been modified to use res_search.  
Also, dn_skip takes an additional parameter and has been renamed
to dn_skipname.  While old sendmail binaries will work with the new
version of bind, because of these changes, it is desirable to install
new sendmail sources and recompile sendmail.  Do not rebuild sendmail 
from old sources.   The new sendmail is on ucbarpa.Berkeley.EDU for
anonymous FTP from pub/4.3/sendmail.MX.tar and pub/4.3/sendmail.MX.tar.Z.

There have been numerous changes to named, fixing most of the known
bugs that can be fixed without major structural changes in the server.
Several server configurations that failed before should now work.
Certain robustness problems have been fixed, in particular bounds-
checking when processing incoming packets.  Two changes have been made
in preparation for negative caching: SOA records are sent in the authority
section in negative responses with NXDOMAIN set, and a bug was fixed that
caused confusion and repeated requests if a response had no error, no answer
and an SOA in the authority section.  As such responses are already sent
by other servers, and will be sent by the next release of BIND, it is
important that all sites upgrade to this version as quickly as possible.

The root "hint" cache and cache file remain the largest problem area,
along with named's naivete in accepting bogus server's data.
These will be addressed in the next release, along with asynchronous
zone transfers, intelligent reloading of zone files, faster startup,
and caching of negative responses.

This version (4.8) will replace the last officially released version (4.5).
Version 4.5 has a serious bug that causes the generation of a continuous
stream of bogons to the root domain servers (bogus queries with the query
response bit set and possibly garbage for nsid and rcode).  It is imperative
that these versions of named be replaced as fast as possible.  We urge you to
field 4.8 quickly, for the sake of the root domain servers.

	Mike Karels
	Jean Wood
	bind@ucbarpa.Berkeley.EDU


usr/sys/dmr/rp.c
#
/*
 *	Copyright 1973 Bell Telephone Laboratories Inc
 */

/*
 * RP disk driver
 */

#include "../param.h"
#include "../buf.h"
#include "../conf.h"
#include "../user.h"

struct {
	int	rpds;
	int	rper;
	int	rpcs;
	int	rpwc;
	int	rpba;
	int	rpca;
	int	rpda;
};
struct { char lbyte, hbyte; };

#define	RPADDR	0176710
#define	NRP	8

struct {
	char	*nblocks;
	int	cyloff;
} rp_sizes[] {
	40600,	0,		/* cyl 0 thru 202 */
	40600,	203,		/* cyl 203 thru 405 */
	9200,	0,		/* cyl 0 thru 45 */
	9200,	360,		/* cyl 360 thru 405 */
	-1,	0,		/* cyl 0 thru 327 */
	-1,	78,		/* cyl 78 thru 405 */
	0,	0,		/* cyl X thru Y */
	0,	0,		/* cyl X thru Y */
};

struct	devtab	rptab;
struct	buf	rrpbuf;

#define	GO	01
#define	RESET	0
#define	RCOM	02
#define	WCOM	04
#define	HSEEK	014

#define	IENABLE	0100
#define	READY	0200

#define	SUFU	01000
#define	SUSU	02000
#define	SUSI	04000
#define	HNF	010000

/*
 * Use av_back to save track+sector,
 * b_resid for cylinder.
 */

#define	trksec	av_back
#define	cylin	b_resid

rpstrategy(abp)
struct buf *abp;
{
	register struct buf *bp;
	register char *p1, *p2;

	bp = abp;
	p1 = &rp_sizes[bp->b_dev.d_minor&07];
	if (bp->b_dev.d_minor >= (NRP<<3) ||
	    bp->b_blkno >= p1->nblocks) {
		bp->b_flags =| B_ERROR;
		iodone(bp);
		return;
	}
	bp->av_forw = 0;
	bp->cylin = p1->cyloff;
	p1 = bp->b_blkno;
	p2 = lrem(p1, 10);
	p1 = ldiv(p1, 10);
	bp->trksec = (p1%20)<<8 | p2;
	bp->cylin =+ p1/20;
	spl5();
	if ((p1 = rptab.d_actf)==0)
		rptab.d_actf = bp;
	else {
		for (; p2 = p1->av_forw; p1 = p2) {
			if (p1->cylin <= bp->cylin
			 && bp->cylin <  p2->cylin
			 || p1->cylin >= bp->cylin
			 && bp->cylin >  p2->cylin) 
				break;
		}
		bp->av_forw = p2;
		p1->av_forw = bp;
	}
	if (rptab.d_active==0)
		rpstart();
	spl0();
}

rpstart()
{
	register struct buf *bp;

	if ((bp = rptab.d_actf) == 0)
		return;
	rptab.d_active++;
	RPADDR->rpda = bp->trksec;
	devstart(bp, &RPADDR->rpca, bp->cylin, bp->b_dev.d_minor>>3);
}

rpintr()
{
	register struct buf *bp;
	register int ctr;

	if (rptab.d_active == 0)
		return;
	bp = rptab.d_actf;
	rptab.d_active = 0;
	if (RPADDR->rpcs < 0) {		/* error bit */
		deverror(bp, RPADDR->rper);
		if(RPADDR->rpds & (SUFU|SUSI|HNF)) {
			RPADDR->rpcs.lbyte = HSEEK|GO;
			ctr = 0;
			while ((RPADDR->rpds&SUSU) && --ctr);
		}
		RPADDR->rpcs = RESET|GO;
		ctr = 0;
		while ((RPADDR->rpcs&READY) == 0 && --ctr);
		if (++rptab.d_errcnt <= 10) {
			rpstart();
			return;
		}
		bp->b_flags =| B_ERROR;
	}
	rptab.d_errcnt = 0;
	rptab.d_actf = bp->av_forw;
	bp->b_resid = RPADDR->rpwc;
	iodone(bp);
	rpstart();
}

rpread(dev)
{

	if(rpphys(dev))
	physio(rpstrategy, &rrpbuf, dev, B_READ);
}

rpwrite(dev)
{

	if(rpphys(dev))
	physio(rpstrategy, &rrpbuf, dev, B_WRITE);
}

rpphys(dev)
{
	register c;

	c = lshift(u.u_offset, -9);
	c =+ ldiv(u.u_count+511, 512);
	if(c > rp_sizes[dev.d_minor & 07].nblocks) {
		u.u_error = ENXIO;
		return(0);
	}
	return(1);
}
